### ISSUE TITLE
IFC105 possible false negative

### ISSUE THREAD

**Issue by RickBrice** [buildingSMART/ifc-gherkin-rules#356]
I'm getting a violation of IFC105, but I don't understand why. The coding seems correct per [4.1.2.6.5 Material Single](https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/concepts/Object_Association/Material_Association/Material_Single/content.html) and [4.1.6.2.2 Property Sets for Materials](https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/concepts/Object_Definition/Property_Sets/Property_Sets_for_Materials/content.html)

![Image](https://github.com/user-attachments/assets/5878242c-ecaa-4a04-b2cb-5537400c0c74)

Here is the IFC snippet

~~~
#859=IFCPROPERTYSINGLEVALUE('CompressiveStrength',$,IFCPRESSUREMEASURE(5.9),#858);
#860=IFCDIMENSIONALEXPONENTS(1,0,0,0,0,0,0);
#861=IFCSIUNIT(*,.LENGTHUNIT.,$,.METRE.);
#862=IFCMEASUREWITHUNIT(IFCLENGTHMEASURE(0.0254),#861);
#863=IFCCONVERSIONBASEDUNIT(#860,.LENGTHUNIT.,'inch',#862);
#864=IFCPROPERTYSINGLEVALUE('MaxAggregateSize',$,IFCPOSITIVELENGTHMEASURE(0.75),#863);
#865=IFCMATERIAL('Precast Concrete, f''c = 5.900 KSI',$,'concrete');
#866=IFCMATERIALPROPERTIES('Pset_MaterialConcrete',$,(#859,#864),#865);
#867=IFCSHAPEREPRESENTATION(#102,'Body','AdvancedSweptSolid',(#845));
#868=IFCPRODUCTDEFINITIONSHAPE($,$,(#867));
#869=IFCBEAM('175Q29X0939esviFnGy5yJ',$,'Segment 1',$,$,#854,#868,$,$);
#870=IFCRELASSOCIATESMATERIAL('3e2K92RorF8xwVZ_4OQbaV',$,'Associates_Concrete_to_Precast_Segment',$,(#869),#865);
~~~

#866 references #865. #865 is referenced by #870 which is a rooted entity.

This follows a similar coding as shown [Example E.8.1](https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/annex_e/structural-analysis-model/structural-curve-member.html)

![Image](https://github.com/user-attachments/assets/6a9ea538-03d7-42d9-86fd-00dfc866d994)

Complete file is attached.

[PGSuper1.ifc.txt](https://github.com/user-attachments/files/18791592/PGSuper1.ifc.txt)

**Comment by aothms** [comment]
We should pay you for doing our job.

We kind of new this though would happen though for IFC105 because the way we implemented it only allows a curated set of whitelisted inverse attributes, because adding all would be too permissive. Because certain things like a shaperepresentation, when it's assigned to a context, it's reachable from project -> context --inv--> representation. But in my eyes that doesn't qualify as that doesn't really mean the shape representation is also used. Quite a big oversight though to not have considered material/profile properties. Quite a few of these novel or compatibility constructs (like properties for materials, where in fact materials should have een object definitions, but that change could not be made due to compatibility) are linked solely with inverse attributes because then it doesn't affect attribute order of preexisting constructs.

**Comment by RickBrice** [comment]
Attached is my original file with 403 errors. They all seem similar to the one I posted yesterday.

[PGSuper1-IFC105-Errors.ifc.txt](https://github.com/user-attachments/files/18800589/PGSuper1-IFC105-Errors.ifc.txt)

**Comment by aothms** [comment]
Thanks Rick that helps, after fixing on our side there are quite a bunch remaining that appear to be truly unreferenced. Mostly axis 2 placements and a couple of swept disks (I only had a cursory look) but this is the list.

```
#14464
#14458
#14459
#14460
#14461
#14462
#14463
#14367
#14368
#14369
#14370
#14371
#14372
#14373
#14029
#14030
#14031
#14032
#14033
#14034
#14035
#14060
#14061
#14062
#14063
#14064
#14065
#14066
#13866
#13867
#13868
#13869
#13870
#13871
#13872
#13223
#13224
#13225
#13226
#13227
#13228
#13229
#13017
#13018
#13019
#13020
#13021
#13022
#13023
#13048
#13049
#13050
#13051
#13052
#13053
#13054
#12710
#12711
#12712
#12713
#12714
#12715
#12716
#12619
#12620
#12621
#12622
#12623
#12624
#12625
#12576
#12577
#12578
#12579
#12580
#12581
#12582
#12583
#12584
#12585
#12586
#12587
#12588
#12589
#12590
#12591
#12592
#12593
#12594
#12538
#12539
#12540
#12541
#11749
#11750
#11751
#11752
#11753
#11754
#11755
#11658
#11659
#11660
#11661
#11662
#11663
#11664
#11351
#11352
#11353
#11354
#11355
#11356
#11357
#11320
#11321
#11322
#11323
#11324
#11325
#11326
#11157
#11158
#11159
#11160
#11161
#11162
#11163
#10514
#10515
#10516
#10517
#10518
#10519
#10520
#10308
#10309
#10310
#10311
#10312
#10313
#10314
#10339
#10340
#10341
#10342
#10343
#10344
#10345
#10001
#10002
#10003
#10004
#10005
#10006
#10007
#9867
#9868
#9869
#9870
#9871
#9872
#9873
#9874
#9875
#9876
#9877
#9878
#9879
#9880
#9881
#9882
#9883
#9884
#9885
#9910
#9911
#9912
#9913
#9914
#9915
#9916
#9829
#9830
#9831
#9832
#9040
#9041
#9042
#9043
#9044
#9045
#9046
#8949
#8950
#8951
#8952
#8953
#8954
#8955
#8642
#8643
#8644
#8645
#8646
#8647
#8648
#8611
#8612
#8613
#8614
#8615
#8616
#8617
#8448
#8449
#8450
#8451
#8452
#8453
#8454
#7808
#7809
#7810
#7811
#7805
#7806
#7807
#7630
#7631
#7632
#7633
#7634
#7635
#7636
#7599
#7600
#7601
#7602
#7603
#7604
#7605
#7296
#7297
#7298
#7292
#7293
#7294
#7295
#7168
#7169
#7170
#7171
#7172
#7173
#7174
#7175
#7176
#7201
#7202
#7203
#7204
#7205
#7206
#7207
#7120
#7121
#7122
#7123
#7158
#7159
#7160
#7161
#7162
#7163
#7164
#7165
#7166
#7167
#6336
#6337
#6331
#6332
#6333
#6334
#6335
#6240
#6241
#6242
#6243
#6244
#6245
#6246
#5902
#5903
#5904
#5905
#5906
#5907
#5908
#5933
#5934
#5935
#5936
#5937
#5938
#5939
#5739
#5740
#5741
#5742
#5743
#5744
#5745
#5096
#5097
#5098
#5099
#5100
#5101
#5102
#4890
#4891
#4892
#4893
#4894
#4895
#4896
#4921
#4922
#4923
#4924
#4925
#4926
#4927
#4583
#4584
#4585
#4586
#4587
#4588
#4589
#4492
#4493
#4494
#4495
#4496
#4497
#4498
#4449
#4450
#4451
#4452
#4453
#4454
#4455
#4456
#4457
#4458
#4459
#4460
#4461
#4462
#4463
#4464
#4465
#4466
#4467
#4411
#4412
#4413
#339
```

**Comment by RickBrice** [comment]
Thanks - not surprised that some of them are on my side. VS is a great IFC learning resource!

**Comment by RickBrice** [comment]
I think I've got them all fixed on my end. The only errors remaining are related to materials. Can you let me know if there are more issues on my side? I presume it will be a few days before you patch the production code.

[PGSuper1.ifc.txt](https://github.com/user-attachments/files/18804169/PGSuper1.ifc.txt)

**Comment by aothms** [comment]
No further errors :)

```
python -m behave --no-capture .\features\IFC105_Resource-entities-need-to-be-referenced-by-rooted-entity.feature --define input=PGSuper1_0215.ifc
@IFC @version2 @E00030 @implementer-agreement
Feature: IFC105 - Resource entities need to be referenced by rooted entity # features/IFC105_Resource-entities-need-to-be-referenced-by-rooted-entity.feature:5
  The rule verifies that resource entities are directly or indirectly related to at least one rooted entity instance by means of forward or a small curated set of inverse attributes.
  Resource entities are the schema classes that do not inherit from IfcRoot, typically defined in the resource layer of the schema (e.g Geometry Resource).
  The inverse attributes that are followed are: StyledByItem HasCoordinateOperation LayerAssignments LayerAssignment HasSubContexts HasProperties (material and profile def) HasRepresentation (material)
  Scenario: Resource entities need to be referenced by rooted entity                                # features/IFC105_Resource-entities-need-to-be-referenced-by-rooted-entity.feature:11
    Given a traversal over the full model originating from subtypes of IfcRoot                      # features/steps/givens/attributes.py:151
    Given an entity instance                                                                        # features/steps/givens/entities.py:18
    Given its entity type is not 'IfcRoot'                                                          # features/steps/givens/attributes.py:65
    Then it must be referenced by an entity instance inheriting from IfcRoot directly or indirectly # features/steps/thens/relations.py:394

1 feature passed, 0 failed, 0 skipped
1 scenario passed, 0 failed, 0 skipped
4 steps passed, 0 failed, 0 skipped, 0 undefined
Took 0m1.137s
```

**Comment by Ghesselink** [comment]
Thomas [implemented a fix](https://github.com/buildingSMART/ifc-gherkin-rules/pull/357), after which the original files pass all tests :-) 
It will indeed take a couple of days for it to be visible on production.

![Image](https://github.com/user-attachments/assets/437d9b97-cf2c-4c7a-910e-be4c93fa762c)

This is the output of the second uploaded file 
[PGSuper1-IFC105-Errors.ifc.txt](https://github.com/user-attachments/files/18800589/PGSuper1-IFC105-Errors.ifc.txt)

![Image](https://github.com/user-attachments/assets/17e11c75-b425-4cdc-bac6-6dfceb4fd616)

As Thomas mentioned, there are also no other errors (also not for the syntax, schema en best-pracitces checks)

**Comment by RickBrice** [comment]
Awesome- thanks. I won’t spend any more time trying to “fix” my file. This IFC105 test defined helps with model quality. There were errors in my code that I probably would have never found without this. Nicely done!
